### PR TITLE
*: use shutdown to close snapshot conn gracefully.

### PR DESCRIPTION
**Pull request by siddontang**
We can use shutdown write first after receiving the whole snapshot, see http://www.gnu.org/software/libc/manual/html_node/Closing-a-Socket.html .

@ngaut @BusyJay @disksing @tiancaiamao

### CODE REVIEW

**Comment by ngaut** [comment]
LGTM

File: src/server/server.rs [hunk 1]
```diff
@@ -204,13 +206,23 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver> Server<T, S> {
...
                     error!("send raft message for token {:?} with msg id {} err {:?}",
                            token,
                            msg_id,
                            e);
                 }
+                if msg_type == RaftMessageType::MsgSnapshot {
+                    // Now we use new connection for every snapshot sending,
+                    // but it is not safe to close the connection after sending OK,
+                    // so we close it here after receiving whole snapshot.
+                    // We use shutdown here for a graceful close.
```
Comment (BusyJay) [inline src/server/server.rs]: How can we verify that the connection is actually closed in the end?

Comment (siddontang) [inline src/server/server.rs]: tcp socket guarantee it. 
After shutdown, the server will send the client FIN, the client knows all data sent to server successfully, then close the socket. Then the server will read 0 later and knows the socket is closed.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**